flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -151,7 +151,12 @@ public void recycle(MemorySegment segment) {
 
 			this.numTotalRequiredBuffers += numRequiredBuffers;
 
-			redistributeBuffers();
+			try {
+				redistributeBuffers();
+			} catch (Throwable t) {
+				this.numTotalRequiredBuffers -= numRequiredBuffers;
+				ExceptionUtils.rethrowIOException(t);
```

Comment:
Shouldn't we also do a clean up of the `LocalBufferPool` in `createBufferPool` if the `redistributeBuffers` fails there? 
```
try {
    redistributeBuffers();
} catch (IOException e) {
    destroyBufferPool(localBufferPool);
    ExceptionUtils.rethrowIOException(e);
}

return localBufferPool;
```

Reply:
After an offline discussion with Nico, we agreed that we should add this bit and additionally also call `redistributeBuffers` in the failure case in `requestMemorySegments`.